### PR TITLE
Option -limitunconfdepth to reject transactions dependent on a long chain of unconfirmed ones

### PR DESCRIPTION
**Pull request by luke-jr**
Not sure if there's any long-term benefit to this policy. Thoughts?

### CODE REVIEW

**Comment by petertodd** [comment]
ACK concept.

Long unconfirmed chains are inefficient - use RBF to replace txs with ones with more outputs - and post problems for future scalability improvements as they reduce parallelism opportunities in block processing.

I'd implement it differently though as an additional depth value recorded for each tx in the mempool - nMaxDepth - where nMaxDepth = max(nMaxDepth of all parent txs) This is efficient to calculate, requiring only a single O(n) pass every time a block is found. There's lots of other use-cases for that kind of architecture too.

**Comment by wizkid057** [comment]
ACK concept and functionality.

I've been testing this commit since last night and it seems to work as intended.

File: src/main.cpp [hunk 1]
```diff
@@ -745,6 +745,32 @@ int64 GetMinFee(const CTransaction& tx, bool fAllowFree, enum GetMinFee_mode mod
 }
 
 
+static
```
Comment (Diapolo) [inline src/main.cpp]: How often do I need to tell you to NOT do this in 2 lines ;)? @laanwj Also did this...

File: src/main.cpp [hunk 2]
```diff
@@ -745,6 +745,32 @@ int64 GetMinFee(const CTransaction& tx, bool fAllowFree, enum GetMinFee_mode mod
...
+unsigned int TxSearchForLongChain(const CTransaction& tx, const CTxMemPool& pool, unsigned int nDepthLimit, unsigned int nDepth = 0)
+{
+    unsigned int nMaxDepth = nDepth;
+    for (unsigned int i = 0; i < tx.vin.size(); ++i)
```
Comment (Diapolo) [inline src/main.cpp]: Nit: ``size_t``?

File: src/main.cpp [hunk 3]
```diff
@@ -798,6 +824,14 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
             break;
         }
     }
+
+        const int nDepthLimit = GetArg("-limitunconfdepth", 0);
+        if (nDepthLimit > 0) {
+            const unsigned int nMaxDepth = TxSearchForLongChain(tx, pool, nDepthLimit);
+            if (nMaxDepth >= (unsigned int)nDepthLimit) {
```
Comment (Diapolo) [inline src/main.cpp]: Couldn't you directly save from GetArg as unsigned int or cast there?

File: src/main.cpp [hunk 4]
```diff
@@ -798,6 +824,14 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
             break;
         }
     }
+
+        const int nDepthLimit = GetArg("-limitunconfdepth", 0);
```
Comment (Diapolo) [inline src/main.cpp]: Is it intended to be not mentioned in the help message?

**Comment by btcdrak** [comment]
Concept ACK

**Comment by matthieu** [comment]
Just worried this patch will increase the incentive to have output spam (especially in light of the current attack). Shred outputs on each block and spend them with more transactions until you hit dust limit and repeat.

**Comment by bencxr** [comment]
NACK, this will exacerbate the effect of full blocks / long confirmation wait times on legitimate users. It impacts scenarios where the wallet software must send out several transactions without knowing the recipients in advance, e.g. users with one unspent sending to multiple payees, exchange wallets that start with one/few unspents (e.g. after refilling a hot wallet). Full RBF to replace txs with more outputs is not a solution today. Further, not sure if the benefits are clear, spammers can send to more outputs, etc.

**Comment by luke-jr** [comment]
@bencxr It won't do anything unless explicitly enabled. Finding the ideal limit to use is an exercise for relay/miner nodes.

**Comment by bencxr** [comment]
@luke-jr what limit would you recommend? 10 - 20? 
many relay/miner node operators do not have the resources to determine the best limit or understand implications.

File: src/main.cpp (hunk 3 above)
Comment (luke-jr) [inline src/main.cpp]: Not until after the non-negative check two lines up...?

File: src/main.cpp (hunk 1 above)
Comment (luke-jr) [inline src/main.cpp]: Sorry, it's a good habit I have...

**Comment by luke-jr** [comment]
@bencxr Part of the reason why it is defaulting to zero is because we do not know what limit to recommend yet. It will take some experimentation. Furthermore, determining proper policy is part of a miner's *job*, so that excuse is not relevant.

**Comment by matthieu** [comment]
@luke-jr there doesn't seem to be a lot of consensus (heh) around what miner's job is. And telling people what their job is rarely helps in making them do it (believe me, I tried).

Furthermore if it's a mining policy, why make it influence pool composition instead or how blocks are built?

**Comment by luke-jr** [comment]
@matthieu It's also a relay policy. I'm just not pushing relay nodes to "do their job" - if they relay spam, it mostly just hurts themselves. :p

**Comment by matthieu** [comment]
@lukejr There are use cases that involve long chains that aren't spammy.
One is people using fee addresses for example, where a service pays for
fees for their users. And is said service can use multiple addresses to
avoid the chain, so can a spammer.

Not questioning that the patch can make sense under a certain light, just
that this specific criteria is one that can help identifying spam.

**Comment by Mirobit** [comment]
@luke-jr You know that most of the node operators and miners will never change a single setting. For node operators that is ok since they do it for free and maybe don't have the time/knowledge to follow all the recent developments. But even miners relay mostly an default policy (e.g. the block size increase from from 0.75 to 1 MB took quite a while and a lot of pushing).

@petertodd If this setting is on default disabled, the miner who enables it is the one that needs more time for block processing. If the miner receives a block from a miner that hasn't enabled it or chose a high number for n, he will not have these tx in the mempool. Thus block verification takes longer.

Thats why default shouldn't be 0 on default. Why not set 1 as default (works with CPFP).

**Comment by jgarzik** [comment]
concept ACK, and I agree rolling this out default-disabled is reasonable.

Main reservation:  Enabling this can potentially create an expensive search, adding CPU cost for a large mempool during new-TX-entry.

**Comment by MarcoFalke** [comment]
Can someone explain me what problem is being solved by this PR? The only one I could think of is "weak" nodes (not miners) trying to be less susceptible to DOS. But then the attacker switches to multi-output txs instead of long tx chains ...

Also, triggering this with ```sendtoaddress``` gives me conflicting transactions. Is this wanted?

**Comment by jgarzik** [comment]
A roll of the dice.

Best to put this in the category of "don't unnecessarily restrict usage" because there may be some edge case utility we don't see immediately.

Weak NAK based on that reasoning.

**Comment by sipa** [comment]
#6554 has better reasoning for this, and should be far more efficient.

**Comment by dcousens** [comment]
NACK.

The DDoS caused by an unlimited depth can easily be replicated by an unlimited horizontal scaling factor.

Consider the case of an attacker choosing to submit 1000 transactions as a form of spam.
In the original case,  he could submit 1 transaction following another, and achieve the desired result.  This would have a depth of 1000.

In the case of a `-limitedunconfdepth` of `20`,  the attack would just submit 1 transaction splitting into 50,  and then 20 levels of 1 transaction following another,  again achieving the desired result of 1000 transactions.

At most,  this will hamper real life use cases of chained transactions.

**Comment by sipa** [comment]
Sorry, I mean #6654 (thanks @dcousens).

**Comment by dcousens** [comment]
*edit:* [from IRC] @sipa I think you referred to the wrong issue? 

NACK.

The DDoS caused by an unlimited depth can easily be replicated by an unlimited horizontal scaling factor.

Consider the case of an attacker choosing to submit 1000 transactions as a form of spam.
Before this pull request,  he could submit 1 transaction following another, and achieve the desired result.  This would have a depth of 1000.

In the case of a `-limitedunconfdepth` of `20`,  the attack would just submit 1 transaction splitting into ~52 outputs,  with ~19 subsequent chains of 1 transaction following another,  again achieving the desired result of ~1000 transactions,  while bypassing the `limitedunconfdepth` of `20`.

At most,  this will hamper real life use cases of chained transactions.

**Comment by gmaxwell** [comment]
@dcousens With respect--- horizontal scaling is not the same thing:

* Horizontal scaling is limited by access to spendable outputs. There is no limit to vertical depth.
* Sufficiently deep chains will (likely) never have their later fees paid (esp as few miners do CPFP and any sane CPFP will have a finite horizon)
* Algorithms which consider decedent or ancestor costs will have usually have super-linear cost in depth.

These points might still support fairly deep chain tolerance, but I do not believe it is correct to say that it is equal to horizontal.

**Comment by dcousens** [comment]
> Horizontal scaling is limited by access to spendable outputs. There is no limit to vertical depth.

Vertical depth is limited by spendable outputs too?   I'm not sure understand how this is different,  every transaction input must have a subsequent output (coin bases excepted)?

> Sufficiently deep chains will (likely) never have their later fees paid (esp as few miners do CPFP and any sane CPFP will have a finite horizon)

Thanks to @sipa I'm aware of CPFP,  but could you explain why deep chains would somehow avoid fees?

> Algorithms which consider decedent or ancestor costs will have usually have super-linear cost in depth.

Could you explain why?

> These points might still support fairly deep chain tolerance, but I do not believe it is correct to say that it is equal to horizontal.

Understandably they have differences,  but my point was that,  for the purposes of a DDoS attack,  they would pose a similar threat for almost the same effort?
However,  that assertion may have some missing assumptions (such as above).

**Comment by sipa** [comment]
Read #6654 please. It introduces more iseful limits, which are inspired by
actual performance problems encountered in the mempool limiting work.

**Comment by petertodd** [comment]
NACK in favor of  #6654. (but thanks for getting the ball rolling on this!)

**Comment by MarcoFalke** [comment]
As #6654 is merged. What is the status of this PR?

**Comment by luke-jr** [comment]
Probably makes sense to close this now.

**Pull request closed without merging**